content/topics/kotlin/generics/_index.md

```diff
@@ -6,16 +6,342 @@ prerequisites:
   soft: []
 ---
 As in Java, classes in Kotlin may have type parameters:
-````
-class Box(t: T) {
+```
+class Box<T>(t: T) {
     var value = t
-    }
-````
+}
+```
 In general, to create an instance of such a class, we need to provide the type arguments:
-````
-val box: Box = Box(1)
-````
+```
+val box: Box<Int> = Box<Int>(1)
+```
 But if the parameters may be inferred, e.g. from the constructor arguments or by some other means, one is allowed to omit the type arguments:
-````
-val box = Box(1) // 1 has type Int, so the compiler figures out that we are talking about Box
-````
+```
+val box = Box(1) // 1 has type Int, so the compiler figures out that we are talking about Box<Int>
+
+``` 
+### Variance
+
+One of the most tricky parts of Java's type system is wildcard types (see Java Generics FAQ). And Kotlin doesn't have any. Instead, it has two other things: declaration-site variance and type projections.
+
+First, let's think about why Java needs those mysterious wildcards. The problem is explained in Effective Java, 3rd Edition, Item 31: Use bounded wildcards to increase API flexibility. First, generic types in Java are invariant, meaning that `List<String>` is not a subtype of `List<Object>`
+
+Why so? If List was not invariant, it would have been no better than Java's arrays, since the following code would have compiled and caused an exception at runtime:
+
+```
+// Java
+List < String > strs = new ArrayList < String >();
+List < Object > objs = strs; // !!! The cause of the upcoming problem sits here. Java prohibits this!
+objs.add(1); // Here we put an Integer into a list of Strings
+String s = strs.get(0); // !!! ClassCastException: Cannot cast Integer to String
+```
+So, Java prohibits such things in order to guarantee run-time safety. But this has some implications. For example, consider the addAll() method from Collection interface. What's the signature of this method? Intuitively, we'd put it this way:
+```
+// Java
+interface Collection<E> ... {
+  void addAll(Collection<E> items);
+}
+```
+But then, we can't do the following simple thing (which is perfectly safe):
+```
+// Java
+void copyAll(Collection<Object> to, Collection<String> from) {
+  to.addAll(from);
+  // !!! Would not compile with the naive declaration of addAll:
+  // Collection<String> is not a subtype of Collection<Object>
+}
+```
+(In Java, we learned this lesson the hard way, see Effective Java, 3rd Edition, Item 28: Prefer lists to arrays)
+
+That's why the actual signature of addAll() is the following:
+```
+// Java
+interface Collection<E> ... {
+  void addAll(Collection<? extends E> items);
+}
+```
+The
+ 
+```
```

Comment:
this would probably look better as inline code

Reply:
Changed